build-type: Configure; package `Cabal-syntax-3.8.1.0` conflicts with itself

```
2024-03-16T10:34:46.0274820Z streamly                     > configure
2024-03-16T10:34:46.2895020Z streamly                     > [1 of 3] Compiling Main             ( /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/stack-885afe2bdf7374d4/streamly-0.8.3/Setup.hs, /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/stack-885afe2bdf7374d4/streamly-0.8.3/.stack-work/dist/x86_64-osx/ghc-9.4.8/setup/Main.o )
2024-03-16T10:34:46.4716440Z streamly                     > [2 of 3] Compiling StackSetupShim   ( /Users/runner/.stack/setup-exe-src/setup-shim-6HauvNHV.hs, /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/stack-885afe2bdf7374d4/streamly-0.8.3/.stack-work/dist/x86_64-osx/ghc-9.4.8/setup/StackSetupShim.o )
2024-03-16T10:34:46.4904380Z streamly                     > 
2024-03-16T10:34:46.4921290Z streamly                     > /Users/runner/.stack/setup-exe-src/setup-shim-6HauvNHV.hs:7:1: error:
2024-03-16T10:34:46.4940030Z streamly                     >     Ambiguous module name ‘Distribution.PackageDescription’:
2024-03-16T10:34:46.5010610Z streamly                     >       it was found in multiple packages:
2024-03-16T10:34:46.5072480Z streamly                     >       Cabal-syntax-3.8.1.0 Cabal-syntax-3.8.1.0
2024-03-16T10:34:46.5082300Z streamly                     >   |
2024-03-16T10:34:46.5123920Z streamly                     > 7 | import Distribution.PackageDescription (PackageDescription, emptyHookedBuildInfo)
2024-03-16T10:34:46.5212090Z streamly                     >   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

https://github.com/haskell/ghcup-hs/actions/runs/8307024100/job/22735795982

stack version 2.15.3

To reproduce:

* `git clone https://github.com/haskell/ghcup-hs.git`
* `stack build`

This seems to only happen on macOS (in CI).

That (the explanation for intermittent failure) seems to be it: locally, if I delete the snapshot package database and start afresh with the example, but in two steps:
~~~text
stack build Cabal-syntax
stack build streamly
~~~

the build fails (with dependency `network`) with:
~~~text
network            > configure
network            > [1 of 3] Compiling Main             ( /private/var/folders/zp/dkvy_dtj31x04x3hcc29wp_00000gn/T/stack-a76fc1f394248344/network-3.1.4.0/Setup.hs, /private/var/folders/zp/dkvy_dtj31x04x3hcc29wp_00000gn/T/stack-a76fc1f394248344/network-3.1.4.0/.stack-work/dist/aarch64-osx/ghc-9.4.8/setup/Main.o )
network            > [2 of 3] Compiling StackSetupShim   ( /Users/mpilgrem/.stack/setup-exe-src/setup-shim-6HauvNHV.hs, /private/var/folders/zp/dkvy_dtj31x04x3hcc29wp_00000gn/T/stack-a76fc1f394248344/network-3.1.4.0/.stack-work/dist/aarch64-osx/ghc-9.4.8/setup/StackSetupShim.o )
network            >
network            > /Users/mpilgrem/.stack/setup-exe-src/setup-shim-6HauvNHV.hs:7:1: error:
network            >     Ambiguous module name ‘Distribution.PackageDescription’:
network            >       it was found in multiple packages:
network            >       Cabal-syntax-3.8.1.0 Cabal-syntax-3.8.1.0
network            >   |
network            > 7 | import Distribution.PackageDescription (PackageDescription, emptyHookedBuildInfo)
network            >   |
~~~

--- a/.stan.toml
+++ b/.stan.toml
@@ -72,14 +72,14 @@
 
 # Anti-pattern: Data.ByteString.Char8.pack
 [[ignore]]
-  id = "OBS-STAN-0203-erw24B-1024:3"
+  id = "OBS-STAN-0203-erw24B-1034:3"
 # ✦ Description:   Usage of 'pack' function that doesn't handle Unicode characters
 # ✦ Category:      #AntiPattern
 # ✦ File:          src\Stack\Build\ExecuteEnv.hs
 #
-#  1023 ┃
-#  1024 ┃   S8.pack . formatTime defaultTimeLocale "%Y-%m-%dT%H:%M:%S%6Q"
-#  1025 ┃   ^^^^^^^
+#  1033 ┃
+#  1034 ┃   S8.pack . formatTime defaultTimeLocale "%Y-%m-%dT%H:%M:%S%6Q"
+#  1035 ┃   ^^^^^^^
 
 # Anti-pattern: Data.ByteString.Char8.pack
 [[ignore]]

--- a/ChangeLog.md
+++ b/ChangeLog.md
@@ -26,6 +26,8 @@ Behaviour changes:
 * The `haddock --haddock-for-hackage` command only seeks to create an archive of
   the `<package_version>-docs` directory for build targets and if flags
   excluding the building of project packages are not set.
+* Stack no longer includes the snapshot package database when compiling the
+  setup executable for a package with `build-type: Configure`.
 
 Other enhancements:
 

--- a/src/Stack/Build/ExecuteEnv.hs
+++ b/src/Stack/Build/ExecuteEnv.hs
@@ -793,20 +793,30 @@ withSingleContext
               -- explicitly requested in the stack.yaml file.
               Nothing -> do
                 warnCustomNoDeps
-                pure $
-                     cabalPackageArg
+                let packageDBArgs' = case package.buildType of
+                      -- The Configure build type is very similar to Simple. As
+                      -- such, Stack builds the setup executable in much the
+                      -- same way as it would in the case of Simple.
+                      C.Configure ->
+                        [ "-hide-all-packages"
+                        , "-package base"
+                        ]
                       -- NOTE: This is different from packageDBArgs above in
                       -- that it does not include the local database and does
                       -- not pass in the -hide-all-packages argument
-                  <> (  "-clear-package-db"
-                     :  "-global-package-db"
-                     :  map
-                          (("-package-db=" ++) . toFilePathNoTrailingSep)
-                          ee.baseConfigOpts.extraDBs
-                     <> [    "-package-db="
-                          <> toFilePathNoTrailingSep ee.baseConfigOpts.snapDB
-                        ]
-                     )
+                      _ ->
+                           map
+                             (("-package-db=" ++) . toFilePathNoTrailingSep)
+                             ee.baseConfigOpts.extraDBs
+                        <> [    "-package-db="
+                             <> toFilePathNoTrailingSep ee.baseConfigOpts.snapDB
+                           ]
+                pure $
+                     [ "-clear-package-db"
+                     , "-global-package-db"
+                     ]
+                  <> packageDBArgs'
+                  <> cabalPackageArg
 
           setupArgs =
             ("--builddir=" ++ toFilePathNoTrailingSep distRelativeDir') : args

